sdk/canton/community/ledger/ledger-common/src/main/scala/com/digitalasset/canton/ledger/error/groups/CommandExecutionErrors.scala

```diff
@@ -649,29 +649,82 @@ object CommandExecutionErrors extends CommandExecutionErrorGroup {
+    object UpgradeError {
+      @Explanation("Validation fails when trying to upgrade the contract")
+      @Resolution(
+        "Verify that neither the signatories, nor the observers, nor the contract key, nor the key's maintainers have changed"
+      )
+      object ValidationFailed
         extends ErrorCode(
-          id = "INTERPRETATION_UPGRADE_ERROR",
+          id = "INTERPRETATION_UPGRADE_ERROR_VALIDATION_FAILED",
           ErrorCategory.InvalidGivenCurrentSystemStateOther,
         ) {
-      final case class Reject(
-          override val cause: String,
-          err: LfInterpretationError.Upgrade.Error,
-      )(implicit
-          loggingContext: ContextualizedErrorLogger
-      ) extends DamlErrorWithDefiniteAnswer(
-            cause = cause
-          ) {
+        final case class Reject(
+                                 override val cause: String,
+                                 err: LfInterpretationError.Upgrade.ValidationFailed,
+                               )(implicit
+                                 loggingContext: ContextualizedErrorLogger
+                               ) extends DamlErrorWithDefiniteAnswer(
+          cause = cause
+        ) {
 
-        override def resources: Seq[(ErrorResource, String)] =
-          Seq(
-            // Class name is reported to help clients (e.g. daml-script) discriminate the type of upgrade error
-            (ErrorResource.UpgradeErrorType, err.getClass.getSimpleName)
-          )
+          override def resources: Seq[(ErrorResource, String)] =
+            Seq(
+              (ErrorResource.ContractId, err.coid.coid),
+              (ErrorResource.TemplateId, err.srcTemplateId.toString),
+              (ErrorResource.TemplateId, err.dstTemplateId.toString),
+              (ErrorResource.Parties, err.signatories.toString),
+              (ErrorResource.Parties, err.observers.toString),
+            ) ++ err.keyOpt.fold(Seq.empty[(ErrorResource, String)])(key => Seq((ErrorResource.ContractKey, key.globalKey.toString), (ErrorResource.Parties, key.maintainers.toString)))
+        }
+      }
+
+      @Explanation("An optional contract field with a value of Some may not be dropped during downgrading")
+      @Resolution(
+        "Ensure optional contract fields with a value of Some are not dropped"
+      )
+      object DowngradeDropDefinedField
+        extends ErrorCode(
+          id = "INTERPRETATION_UPGRADE_ERROR_DOWNGRADE_DROP_DEFINED_FIELD",
+          ErrorCategory.InvalidGivenCurrentSystemStateOther,
+        ) {
+        final case class Reject(
+                                 override val cause: String,
+                                 err: LfInterpretationError.Upgrade.DowngradeDropDefinedField,
+                               )(implicit
+                                 loggingContext: ContextualizedErrorLogger
+                               ) extends DamlErrorWithDefiniteAnswer(
+          cause = cause
+        )
```

Comment:
Yes. With my original thinking, both fields would involve converting `Value` to a string. My current understanding is that this could lead to grpc error headers being truncated, hence why I'd avoided doing this.

Looking again at the contents of `Error.Upgrade.DowngradeDropDefinedField`, I've realised that the first field has type `Ast.Type` which might be a reasonable value to serialise.

Reply:
Added in changes to capture `expectedType`

Reply:
Is it not possible to simply include the field name?

Reply:
I don't think this is currently possible? The current data structure we have to work with is `DowngradeDropDefinedField(expectedType: Ast.Type, actualValue: Value)` and there's an associated TODO (https://github.com/digital-asset/daml/issues/17647) that likely is work that would enable doing something like this.

Reply:
I think that's worth doing, just extending the error type if its easy.
idm if its this Pr or another

Reply:
I plan to do this as part of the work under UP-20 (in addition to handling the lookup errors).